### ISSUE TITLE
shellcheck.executablePath correct, but getting "shellcheck...not found" error

### ISSUE THREAD

**Issue by bland328** [vscode-shellcheck/vscode-shellcheck#767]
Using VS Code 1.72.2 (Universal) under macOS Monterey, I have `shellcheck` version `0.8.0` installed via `homebrew`, with `shellcheck.executablePath` set to `/opt/homebrew/bin/shellcheck` _(the standard `homebrew` installation path for Apple Silicon-based systems)_.

This has worked fine for months, but recently I started getting this error on VS Code startup:

```
The shellcheck program was not found (not installed?).
Use the 'shellcheck.executablePath' setting to configure the location of 'shellcheck'
```

I'm unclear if this error started appearing after a VS Code update, a `vscode-shellcheck` update, a `shellcheck` update, or something else, though I can try to narrow this, if needed.

**What I've tried to fix this:**

- Uninstalled `vscode-shellcheck`, deleted `~/.vscode/extensions/timonwong.shellcheck-0.23.1` and reinstalled
- Tried `/opt/homebrew/Cellar/shellcheck/0.8.0/bin/shellcheck` instead _(path of the actual binary targeted by the `/opt/homebrew/bin/shellcheck` symlink)_
- Copy/pasted path into terminal to confirm correct
- Confirmed permissions of `shellcheck` binary as `0555`

_**Edit:** I originally posted that `vscode-shellcheck` was working despite the error, but that wasn't accurate; I was apparently seeing cached results for tabs that had been open since before the error first appeared._

**Comment by bland328** [comment]
FYI, I just tried removing the `shellcheck.executablePath` in an attempt to use the built-in `shellcheck` binary, but the error remains.

**Comment by bland328** [comment]
I've discovered the `*ShellCheck.log` files under `~/Library/Application\ Support/Code/logs/*`, and they contain various combinations of:
```
[INFO] shellcheck version: 0.8.0
```
and
```
[INFO] shellcheck (bundled) version: 0.8.0
```
suggesting to me that the `shellcheck` executables are, in fact, being found.

If there's a way to boost the logging detail, I'm happy to collect more information.

**Comment by felipecrs** [comment]
Interesting. Are you sure your settings are completely empty? For example, you may have workspace-level settings.json overriding the executablePath.

**Comment by timonwong** [comment]
Just FYI, there is an indicator if settings are overrided somewhere else:

Default settings, modified in workspace:
<img width="509" alt="image" src="https://user-images.githubusercontent.com/1398228/196076857-7a28f55f-b373-4d22-9d46-b3bc212f032a.png">

Changed user settings, also modified in workspace:
<img width="528" alt="image" src="https://user-images.githubusercontent.com/1398228/196076913-cdcf63f4-c204-4b62-b9c1-05ac20e6e011.png">

**Comment by bland328** [comment]
> Are you sure your settings are completely empty?

Yes. I've configured the executable path as a `User` setting, and under both the `Workspace` tab and the tab for my current project folder, I see _`(Modified in User)`_.

Plus, just to be thorough, I checked the various JSON settings files, and found nothing except for
```
"shellcheck.executablePath": "/opt/homebrew/Cellar/shellcheck/0.8.0/bin/shellcheck",
```
in the User Settings `settings.json` file.

Happy to check whatever else might help.

**Comment by felipecrs** [comment]
Have you tried unsetting the `executablePath`, so that the extension will use the bundled one?

**Comment by timonwong** [comment]
#697 reminds me now the settings may come from many sources... so executablePath will be also override-able in folder, language specific settings, etc...

**Comment by timonwong** [comment]
@bland328 In #774 we add a new option `shellcheck.logLevel`, if you set it to "debug" or "trace" it will give you more information in log (in output-> Shellcheck)
<img width="1018" alt="image" src="https://user-images.githubusercontent.com/1398228/196435797-5c5e9ad1-fd26-4113-9de9-98756fc82ced.png">

You can download the vsix [here](https://github.com/vscode-shellcheck/vscode-shellcheck/suites/8828996412/artifacts/402197244): https://github.com/vscode-shellcheck/vscode-shellcheck/actions/runs/3272503666

**Comment by bland328** [comment]
@timonwong, I tried the newer vsix with `shellcheck.logLevel` set to `trace`.

Here's the output with `"shellcheck.executablePath": "/opt/homebrew/Cellar/shellcheck/0.8.0/bin/shellcheck"`:
```
[INFO]: shellcheck version: 0.8.0
[DEBUG]: Spawn: /opt/homebrew/Cellar/shellcheck/0.8.0/bin/shellcheck -f json1 -
[DEBUG]: Unable to start shellcheck: [Error: spawn /opt/homebrew/Cellar/shellcheck/0.8.0/bin/shellcheck ENOENT
	at ChildProcess._handle.onexit (node:internal/child_process:283:19)
	at onErrorNT (node:internal/child_process:478:16)
	at processTicksAndRejections (node:internal/process/task_queues:83:21)
	at runNextTicks (node:internal/process/task_queues:65:3)
	at listOnTimeout (node:internal/timers:528:9)
	at process.processTimers (node:internal/timers:502:7)] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawn /opt/homebrew/Cellar/shellcheck/0.8.0/bin/shellcheck',
  path: '/opt/homebrew/Cellar/shellcheck/0.8.0/bin/shellcheck',
  spawnargs: [ '-f', 'json1', '-' ]
}
```
And here's the output with `"shellcheck.executablePath": ""`:
```
[INFO]: shellcheck (bundled) version: 0.8.0
[DEBUG]: Spawn: /Users/brian/.vscode/extensions/timonwong.shellcheck-0.23.2-pr-774.0/binaries/darwin/arm64/shellcheck -f json1 -
[DEBUG]: Unable to start shellcheck: [Error: spawn /Users/brian/.vscode/extensions/timonwong.shellcheck-0.23.2-pr-774.0/binaries/darwin/arm64/shellcheck ENOENT
	at ChildProcess._handle.onexit (node:internal/child_process:283:19)
	at onErrorNT (node:internal/child_process:478:16)
	at processTicksAndRejections (node:internal/process/task_queues:83:21)
	at runNextTicks (node:internal/process/task_queues:65:3)
	at listOnTimeout (node:internal/timers:528:9)
	at process.processTimers (node:internal/timers:502:7)] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawn /Users/brian/.vscode/extensions/timonwong.shellcheck-0.23.2-pr-774.0/binaries/darwin/arm64/shellcheck',
  path: '/Users/brian/.vscode/extensions/timonwong.shellcheck-0.23.2-pr-774.0/binaries/darwin/arm64/shellcheck',
  spawnargs: [ '-f', 'json1', '-' ]
}
```

In both cases, when I combine `path` and `spawnargs` and run from a terminal, `shellcheck` successfully runs _(but does not return until I hit Ctrl-C, which I gather is appropriate because it is awaiting `stdin`, thanks to the trailing hyphen?)_.

**Comment by timonwong** [comment]
Interesting, so it seems the version check is ok (it also spawned the shellcheck binary), but failed with cmd & args for actual linting.

**Comment by timonwong** [comment]
@bland328 Can you post your full vscode version? Menu -> Code -> About Visual Studio Code, then click "Copy"

<img width="404" alt="image" src="https://user-images.githubusercontent.com/1398228/196586864-0697048d-4c4d-4c5f-990a-64653113bca8.png">

For example:

```
Version: 1.72.2
Commit: d045a5eda657f4d7b676dedbfa7aab8207f8a075
Date: 2022-10-12T22:15:55.763Z
Electron: 19.0.17
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Darwin arm64 21.6.0
Sandboxed: No
```

**Comment by timonwong** [comment]
> Interesting, so it seems the version check is ok (it also spawned the shellcheck binary using `shellcheck -V`), but failed with cmd & args for actual linting.

The main difference is the API which called:
- version check: `execa.sync`
- linter: using `execa()` to spawn a child process

**Comment by felipecrs** [comment]
> version check: execa.sync

I can't explain, but I added this to fix a similar issue in the past:
https://github.com/vscode-shellcheck/vscode-shellcheck/pull/166

**Comment by felipecrs** [comment]
Hey, @bland328 you said you are using VS Code universal. Can you try using the Apple Silicon one?

**Comment by timonwong** [comment]
nodejs doc [states that](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options):

> Use `cwd` to specify the working directory from which the process is spawned. If not given, the default is to inherit the current working directory. If given, but the path does not exist, the child process emits an ENOENT error and exits immediately

So maybe cwd is also a problem here?

**Comment by timonwong** [comment]
@bland328 Can you try this https://github.com/vscode-shellcheck/vscode-shellcheck/suites/8869725436/artifacts/405104364 ?

**Comment by bland328** [comment]
@timonwong, it appears to me that the "bad cwd" hypothesis is correct, and that `shellcheck-0.25.1-pr-780.0.vsix` fixes the problem.

This morning, the problem vanished; `vscode-shellcheck` was suddenly back to working fine for me without any further intentional experiments or configuration changes.

So, I did this:
```
$ mkdir /tmp/test
$ code /tmp/test
$ rmdir /tmp/test
```

Then, in VS Code, I opened a new tab and typed `#!/bin/bash`, at which point the problem reappeared.

I then installed `shellcheck-0.25.1-pr-780.0.vsix`, tried my test again from scratch, and had no problem.

If you'd still like for me to post the specifics of my VS Code version and/or test the Apple Silicon build, I'm happy to help. Thanks for the help!
